### PR TITLE
MINIFICPP-2024 Refactor repository hierarchy

### PR DESCRIPTION
**Pull request by lordgamez**
Addresses issues listed in https://issues.apache.org/jira/browse/MINIFICPP-2024

Prerequisite for https://issues.apache.org/jira/browse/MINIFICPP-2022

----------------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by adamdebreceni — APPROVED** [review]

**Review by szaszm — APPROVED** [review]

File: libminifi/src/core/repository/VolatileContentRepository.cpp [hunk 1]
```diff
@@ -45,10 +48,10 @@ bool VolatileContentRepository::initialize(const std::shared_ptr<Configure> &con
     }
   }
   if (!minimize_locking_) {
-    for (auto ent : value_vector_) {
+    for (auto ent : repo_data_.value_vector) {
       delete ent;
     }
-    value_vector_.clear();
+    repo_data_.value_vector.clear();
```
Comment (fgerlits) [inline libminifi/src/core/repository/VolatileContentRepository.cpp:54]: I would move this to `VolatileRepositoryData` as a new `clear()` function.

File: libminifi/src/core/repository/VolatileContentRepository.cpp [hunk 2]
```diff
@@ -86,7 +89,7 @@ std::shared_ptr<io::BaseStream> VolatileContentRepository::write(const minifi::R
     if (claim_check != master_list_.end()) {
       return std::make_shared<io::AtomicEntryStream<ResourceClaim::Path>>(claim.getContentFullPath(), claim_check->second);
     } else {
-      auto *ent = new AtomicEntry<ResourceClaim::Path>(&current_size_, &max_size_);
+      auto *ent = new AtomicEntry<ResourceClaim::Path>(&repo_data_.current_size, &repo_data_.max_size);
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/src/core/repository/VolatileContentRepository.cpp:89]: I'm not sure if it's a good idea as it is stored in the `master_list_` as a raw pointer where the ownership is not trivial. We have the `value_vector` of the repo data and the `master_list_` and the ownership between those 2 needs to be clarified first to see how the ownership should be handled.

File: libminifi/include/core/repository/FileSystemRepository.h [hunk 3]
```diff
@@ -22,39 +22,31 @@
...
-#include "core/Core.h"
 #include "../ContentRepository.h"
 #include "properties/Configure.h"
 #include "core/logging/LoggerFactory.h"
 
 namespace org::apache::nifi::minifi::core::repository {
 
-/**
- * FileSystemRepository is a content repository that stores data onto the local file system.
- */
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/include/core/repository/FileSystemRepository.h:34]: It's true that it is not obvious from the class name, but the definition of the class states that it is inherited from the `ContentRepository` which I think makes it clear that it is a content repository, so I don't think this comment is needed.

File: libminifi/src/core/repository/VolatileRepository.cpp [hunk 4]
```diff
@@ -23,16 +23,144 @@
...
+namespace org::apache::nifi::minifi::core::repository {
+
+void VolatileRepository::loadComponent(const std::shared_ptr<core::ContentRepository>& /*content_repo*/) {
+}
+
+VolatileRepository::~VolatileRepository() {
+  for (auto ent : repo_data_.value_vector) {
+    delete ent;
+  }
+}
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/src/core/repository/VolatileRepository.cpp]: Updated in 79d5a6e5d2d061953ebc87b82d8cc10da6bf10d6

File: libminifi/src/core/repository/VolatileContentRepository.cpp (hunk 1 above)
Comment (lordgamez) [reply libminifi/src/core/repository/VolatileContentRepository.cpp]: Updated in 79d5a6e5d2d061953ebc87b82d8cc10da6bf10d6

File: libminifi/src/core/repository/VolatileRepository.cpp [hunk 5]
```diff
@@ -23,16 +23,141 @@
...
-} /* namespace nifi */
-} /* namespace apache */
-} /* namespace org */
+namespace org::apache::nifi::minifi::core::repository {
+
+void VolatileRepository::loadComponent(const std::shared_ptr<core::ContentRepository>& /*content_repo*/) {
+}
+
+VolatileRepository::~VolatileRepository() {
+}
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/src/core/repository/VolatileRepository.cpp]: Good point, removed in 688c4a816f3d4ecafa721771f8fa2452d9b4eb08

**Review by fgerlits — APPROVED** [review]

File: libminifi/include/core/repository/VolatileRepositoryData.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+#include <memory>
+
+#include "AtomicRepoEntries.h"
+#include "properties/Configure.h"
+
+namespace org::apache::nifi::minifi::core::repository {
+
+struct VolatileRepositoryData {
+  static const char *volatile_repo_max_count;
+  static const char *volatile_repo_max_bytes;
```
Comment (szaszm) [inline libminifi/include/core/repository/VolatileRepositoryData.h:31]: Please make these top-level const

Comment (lordgamez) [reply libminifi/include/core/repository/VolatileRepositoryData.h]: Updated in fb134177002b780f93f09a1c43ebef605cc691e0

**Review by szaszm — APPROVED** [review]

**Pull request closed without merging**